text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,426 @@
+
+``` json
+{
+  "<layer1-diffID>": [
+    {
+      "name": "java11",
+      "digest": "sha256:java11-asset-sha256",
+      "uri": "https://path/to/java11.tgz",
+    },
+    {
+      "digest": "sha256:<java13-asset-sha256>",
+      "metadata": {
+        "name": "java13"
+      }
+    }
+  ],
+  "...": [],
+  "<layern-diffID>": [
+    {
+      "digest": "sha256:<java15-asset-sha256>",
+    }
+  ]
+}
+```
+
+##### Buildpackage Changes
+
+To enable buildpackages to reference asset caches, we define a new buildpackage label, `io.buildpacks.buildpackage.assets`.
+
+The contents of this label will be a `json` with a list of data describing each asset cache. This will allow the platform to make decisions about which assets to pull before a build.
+
+The format of this object will be as follows:
+ 
+A top level `assetCache` field listing possible asset caches. Each entry in the `assetCache` field will list the `name`, the asset image `refDigest`, an alternative `id`, as well as `layersDiffIDs` field with contents identical to the `io.buildpacks.asset.layers`  label on the associated asset cache.
+
+##### Buildpackage Label
+[buildpackage-label]: #buildpackage-label
+``` json
+{
+  "assetCache": [
+    {
+      "ref": "gcr.io/buildpacks/java-asset-cache",
+      "refDigest": "sha256:<some-java-asset-cache-sha256>",
+      "name": "buildpacks-assets/java-assets@1.1.1",
+      "layerDiffIDs": {
+        "<layer1-diffID>": [
+          {
+            "digest": "sha256:java11-asset-sha256",
+            "uri": "https://path/to/java11.tgz",
+            "metadata": {}
+          },
+          {
+            "digest": "sha256:<java13-asset-sha256>",
+            "metadata": {
+              "extra": "java13 metadata"
+            }
+          }
+        ],
+        "<layer2-diffID>": [
+          "..."
+        ]
+      }
+    },
+    {
+      "ref": "https://buildpacks.io/asset-cache-fallback/java.tgz",
+      "refDigest": "sha256:<some-asset-cache-fallback-sha256>",
+      "name": "buildpack-assets/asset-cache-fallback@1.2.3",
+      "layerDiffIDs": {
+        "<other-layer1-diffID>": [
+          {
+            "digest": "sha256:<java15-asset-sha256>",
+            "uri": "/local/path/to/java15.zip",
+            "metadata": {}
+          }
+        ]
+      }
+    }
+  ]
+}
+```
+
+##### Builder Changes
+
+Builders that contain asset caches will have an `io.buildpacks.asset.layers` with the same format as the asset cache label. Again this label will contain a `<layer1-diffID>` entry for each asset layer contained in the builder.
+
+Builders will additionally need to retain the asset references for each buildpackage in the builder. For this we will extend the existing `io.buildpacks.buildpack.layers` label to contain mappings to contain an `cachedAssets` field identical to the one above.
```

Comment:
Why will builders need this information as well. Would it be sufficient for the builder to have a label with information equivalent to: `io.buildpacks.buildpackage.assets`?

Reply:
Ya, I was thinking this would be good information to let platforms have access to allow them to make decisions about pulling asset images before a build. This asset image would then be available locally for all subsequent builds.